mmgen/models/architectures/dalle_mini/modules.py

```diff
@@ -0,0 +1,235 @@
+        self.fc3 = Linear(mid_channels, in_out_channels, bias=False)
+        self.gelu = build_activation_layer(dict(type='GELU'))
+
+    def forward(self, z):
+        """Forward function.
+
+        Args:
+            z (torch.FloatTensor): Input feature map.
+
+        Returns:
+            z (torch.FloatTensor): Output feature map.
+        """
+        z = self.norm1(z)
+        w = self.fc1(z)
+        w = self.gelu(w)
+        v = self.fc2(z)
+        z = self.norm2(w * v)
+        z = self.fc3(z)
+        return z
+
+
+@MODULES.register_module()
+class AttentionBase(nn.Module):
+    """An Muti-head Attention block used in Bart model.
+
+    Ref:
+    https://github.com/kuprel/min-dalle/blob/main/min_dalle/models
+
+    Args:
+        in_channels (int): The channel number of the input feature map.
+        head_num (int): Number of heads in the attention.
+    """
+
+    def __init__(self, in_channels, head_num):
+        super().__init__()
+        self.in_channels = in_channels
+        self.head_num = head_num
+        self.querie = Linear(in_channels, in_channels, bias=False)
+        self.key = Linear(in_channels, in_channels, bias=False)
+        self.value = Linear(in_channels, in_channels, bias=False)
+        self.proj = Linear(in_channels, in_channels, bias=False)
+
+    def qkv(self, x):
+        """Calculate queries, keys and values for the embedding map.
+
+        Args:
+            x (torch.FloatTensor): Input feature map.
+
+        Returns:
+            q (torch.FloatTensor): Querie feature map.
+            k (torch.FloatTensor): Key feature map.
+            v (torch.FloatTensor): Value feature map.
+        """
+        q = self.querie(x)
+        k = self.key(x)
+        v = self.value(x)
+
+        return q, k, v
+
+    def forward(self, q, k, v, attention_mask):
+        """Forward function for attention.
+
+        Args:
+            q (torch.FloatTensor): Querie feature map.
+            k (torch.FloatTensor): Key feature map.
+            v (torch.FloatTensor): Value feature map.
+            attention_mask (torch.BoolTensor): whether to use
+                                                an attention mask.
+
+        Returns:
+            weights (torch.FloatTensor): Feature map after attention.
+        """
+        q = q.reshape(q.shape[:2] + (self.head_num, -1))
+        q /= q.shape[-1]**0.5
+        k = k.reshape(k.shape[:2] + (self.head_num, -1))
+        v = v.reshape(v.shape[:2] + (self.head_num, -1))
+
+        attention_bias = (1 - attention_mask.to(torch.float32)) * -1e12
+        weights = torch.einsum('bqhc,bkhc->bhqk', q, k)
+        weights += attention_bias
+        weights = torch.softmax(weights, -1)
+        weights = torch.einsum('bhqk,bkhc->bqhc', weights, v)
+        shape = weights.shape[:2] + (self.in_channels, )
+        weights = weights.reshape(shape)
+        weights = self.proj(weights)
+        return weights
+
+
+@MODULES.register_module()
+class EncoderLayer(nn.Module):
```

Comment:
Since this encoder layer is used for Bart, `BartEncoderLayer` may be a better name.